crypto_box.js

```diff
@@ -95,6 +104,166 @@ function crypto_box_seal_open (m, c, pk, sk) {
+//                     const unsigned char *sk)
+// {
+//     unsigned char k[crypto_box_BEFORENMBYTES];
+//     int           ret;
+//
+//     COMPILER_ASSERT(crypto_box_BEFORENMBYTES >= crypto_secretbox_KEYBYTES);
+//     if (crypto_box_beforenm(k, pk, sk) != 0) {
+//         return -1;
+//     }
+//     ret = crypto_box_detached_afternm(c, mac, m, mlen, n, k);
+//     sodium_memzero(k, sizeof k);
+//
+//     return ret;
+// }
+function crypto_box_detached (c, mac, m, n, pk, sk) {
+  check(mac, crypto_box_MACBYTES)
+  check(n, crypto_box_NONCEBYTES)
+  check(pk, crypto_box_PUBLICKEYBYTES)
+  check(sk, crypto_box_SECRETKEYBYTES)
+
+  const k = new Uint8Array(crypto_box_BEFORENMBYTES)
+
+  if (crypto_box_beforenm(k, pk, sk) !== 0) {
+    return -1
+  }
+
+  cleanup(k)
+
+  return crypto_box_detached_afternm(c, mac, m, n, k)
+}
+
+// int
+// crypto_box_easy(unsigned char *c, const unsigned char *m,
+//                 unsigned long long mlen, const unsigned char *n,
+//                 const unsigned char *pk, const unsigned char *sk)
+// {
+//     if (mlen > crypto_box_MESSAGEBYTES_MAX) {
+//         sodium_misuse();
+//     }
+//     return crypto_box_detached(c + crypto_box_MACBYTES, c, m, mlen, n,
+//                                pk, sk);
+// }
+function crypto_box_easy (c, m, n, pk, sk) {
+  assert(c.length === m.length + crypto_box_MACBYTES, "c should be a buffer of length 'm.length + crypto_box_MACBYTES'")
+  console.log({ c, m, crypto_box_MACBYTES })
+  console.log('yep')
+  assert(c.byteLength <= crypto_box_MESSAGEBYTES_MAX, "m should not be more than 'crypto_box_MESSAGEBYTES_MAX' bytes")
+
+  return crypto_box_detached(c.subarray(crypto_box_MACBYTES), c.subarray(0, crypto_box_MACBYTES), m, n, pk, sk)
+}
+// int
+// crypto_box_open_detached_afternm(unsigned char *m, const unsigned char *c,
+//                                  const unsigned char *mac,
+//                                  unsigned long long clen,
+//                                  const unsigned char *n,
+//                                  const unsigned char *k)
+// {
+//     return crypto_secretbox_open_detached(m, c, mac, clen, n, k);
+// }
+function crypto_box_open_detached_afternm (m, c, mac, n, k) {
+  return crypto_secretbox_open_detached(m, c, mac, n, k)
+}
+
+// int
+// crypto_box_open_detached(unsigned char *m, const unsigned char *c,
+//                          const unsigned char *mac,
+//                          unsigned long long clen, const unsigned char *n,
+//                          const unsigned char *pk, const unsigned char *sk)
+// {
+//     unsigned char k[crypto_box_BEFORENMBYTES];
+//     int           ret;
+//
+//     if (crypto_box_beforenm(k, pk, sk) != 0) {
+//         return -1;
+//     }
+//     ret = crypto_box_open_detached_afternm(m, c, mac, clen, n, k);
+//     sodium_memzero(k, sizeof k);
+//
+//     return ret;
+// }
+function crypto_box_open_detached (m, c, mac, n, pk, sk) {
+  check(mac, crypto_box_MACBYTES)
```

Comment:
`check` is used in the rest of the module and works here, but just for future reference, we would like to eventually replace `check` with standard `asserts` using nanoassert, so this would become:
`assert(mac.byteLength === crypto_box_MACBYTES, "mac should be 'crypto_box_MACBYTES' in length")`

Constants are called specifically to encourage users for clarity and users should consult the docs to find their values.